man/man1/fitsgids.1
.TH FITSGIDS 1NEMO "13 April 1995"
.SH NAME
fitsgids, mirgids \- display an image on the GIDS display server
.SH SYNOPSIS
\fBfitsgids\fP \fBin=\fP\fIfits_file\fP [parameter=value]
.SH DESCRIPTION
\fBfitsgids\fP loads an image from (the primary HDU of) 
a fitsfile into the \fBgids\fP display
server (see also \fIgids(lNEMO)\fP).
Through the server program the images can then be manipulated, although
there is a selection of a transfer function (see \fBfilter=\fP below).
By default all planes from the input file are loaded,
and GIDS can also be used to display them
in \fImovie\fP mode XY-XZ-YZ \fIslice\fP mode (see EXAMPLES below).
.PP
\fBfitsgids\fP also prepares GIDS automatically for movie sequencing, 
such that the plane numbers do not have to be typed in, one can directly
go to the \fBLoop\fP command. This is however not the case for \fIslice\fP 
mode.
.PP
\fBmirgids\fP expects a miriad dataset instead of a fits file, but
is otherwise the same.
.SH PARAMETERS
The following parameters are recognized in any order if the keyword
is also given:
.TP 20
\fBin=\fP\fIfits_file\fP
Input file. Expected to be in \fIfits(5NEMO)\fP format for 
\fBfitsgids\fP, and a miriad image dataset for \fBmirgids\fP.
No default.
.TP
\fBdatamin=\fP
Datamin, if to override. By default the value of the DATAMIN 
fits header keyword will be used, otherwise no default.
The value applies to the actual physical datavalue, not the filtered value
(see \fBfilter=\fP below). The datamin/max values the program reports at the
end are the filtered values.
.TP
\fBdatamax=\fP
Datamax, if to override. By default the value of the DATAMAX
fits header keyword will be used, otherwise no default.
.TP
\fBx=\fP
List of X pixels to display. This can be any \fInemoinpi(3NEMO)\fP expression,
and does not have to be in order. Must be in range 1..NAXIS1. [all].
.TP
\fBy=\fP
List of Y pixels to display. Valid ranges 1..NAXIS2, see above. [all]. 
.TP
\fBz=\fP
List of Z pixels to display. Valid ranges 1..NAXIS3, see above. [all].
.TP
\fBplane=\fP\fIfirst_plane\fP
Plane to start recording in. Valid planes are 1 to the largest number
your GIDS server allows. Any other number is invalid, and bypasses
calls to the GIDS server. Default: 1.
.TP
\fBfilter=\fP\fItransfer_function\fP
An additional transfer function can be supplied here by which the data
will be ``scaled''. Valid options are \fBlinear, sqrt, sqr, log, exp\fP.
[Default: \fBlinear\fP].
.SH EXAMPLES
The following command loads all planes from a datacube \fIm51.cube.fits\fP 
(assumed to contain 41 planes in this example)
with a stretch from -10 to 90:
.nf
    % \fBfitsgids m51.cube.fits -10 90\fP
.fi
Next, to get into GIDS \fImovie\fP mode,
go through the \fBEtcetera/Recording\fP menus and type in the
planes that you want to see in a movie in the 7th (blank) widget in
the upper left corner of the GIDS server, i.e.
.nf
    \fBEtcetera\fP
    \fBRecording\fP
    \fB1:41\fP    (\fI in the entry window -- not required if auto-sequenced\fP)
    \fBLoop\fP
.fi
will show planes 1 through 41 (press the middle
mouse button on \fBLoop\fP menu for context sensitive help), 
and use the middle mouse
button in the display window to control the speed. Exit from this
with \fBReturn\fP.
.PP
To get into \fIslice\fP mode  (3 simultaneous orthogonal views of 
the datacube) from the previous (movie) setup type:
.nf
    \fBReturn\fP
    \fBSlice\fP
    \fB1:41\fP        (\fI in the entry window\fP)
    \fBXslice\fP
    \fBZoom X\fP
    \fB1:41\fP        (\fI in the entry window\fP)
    \fBYslice\fP
.fi
and point with any mousebutton on features in any of the 3 windows, and
watch the display change.
.SH BUGS
If you do not have the \fIgipsy\fP environment loaded,
and/or the \fBgids\fP server program has not been loaded yet,
\fBfitsgids\fP will attempt to do this for you, 
but may not be able to find the help file. 
To remidy this, start up \fBgids\fP manually, with enough
memory allocated to store images, before. For example:
.nf
    % gids -memory 10000
.fi
would allocate 10,000 Kb. Remember you can always use \fBResize\fP to
a smaller display, and hence allocate more image planes to use for
storing movies. The  \fIgipsy\fP environment should define an environment
variable \fB$gids_setup\fP
.PP
Program needs to be compiled with the \fIlibgdi.a\fP, not GIPSY's
\fPgiplib.a\fP. Various fortran I/O libraries would otherwise be needed
to complete linking.
.SH SEE ALSO
gids(lNEMO), saoimage(1), ds(1), fitsio(3NEMO), nemoinpi(3NEMO), fits(5NEMO)
.SH FILES
.ta +2i
.nf
$gids_setup	directory in which the tvdevices file lives
DEFAULT_DISPLAY	temp file, created by ... contains
src/image/fits	fitsgids.c (main), look_nemo.c (nemo-gipsy interface)
usr/gipsy	gids, libgdi (see also ftp://kapteyn.astro.rug.nl/gids/gids-export.tar.Z)
.fi
.SH AUTHOR
Peter Teuben
.SH UPDATE HISTORY
.nf
.ta +1.0i +4.0i
140may-93	V0.x initial rough cut version written	PJT
21-Jul-94	V1.0 implemented x,y,z= and fixed scaling; added much needed cache	PJT
27-mar-95	documented some of the peculiarities of startup files	PJT
13-apr-95	V1.1 added filter=	PJT
1-feb-98	V1.1a more verbose error messages	PJT
.fi


src/image/fits/fitsgids.c
/*
 *   FITSGIDS:   dump fits file onto the GIDS server
 *
 *      14-may-93   written  (not completed yet)
 *	20-jul-94   implemented the *range= keywords: recording planes
 *      21-jul-94   cashing in the look_ routines for efficiency
 *		    (improved a 512*512 image loading time from 1'43" to 3" !)
 *                  renamed xrange= -> x= etc. and id->record->plane=...    PJT
 *	13-apr-95   add filter (transfer function) before output
 *       1-feb-98   also display the 'gid' error message from xlibc's   PJT
 *                  (see look_nemo.c)
 */

#include <stdinc.h>
#include <getparam.h>
#include <extstring.h>
#include <image.h>
#include <fitsio.h>

string defv[] = {
    "in=???\n		Input fits file",
    "datamin=\n         Datamin, if to override",
    "datamax=\n         Datamax, if to override",
    "x=\n               Range in X to display [all]",
    "y=\n               Range in Y to display [all]",
    "z=\n               Range in Y to display [first]",
    "plane=1\n          Plane to start recording in (0=no record)",
    "filter=linear\n    Filter applied to output",
    "VERSION=1.1a\n	1-feb-98 PJT",
    NULL,
};

string usage = "display fits images on GIDS display server";

#ifndef MAXSIZE
#define MAXSIZE  2048	/* max size in each dimension of an image */
#endif

extern string *burststring(string, string);

local int make_idx(string, int *, int, int);
local real filter_lin(double), filter_sqrt(double), filter_sqr(double),
           filter_log(double), filter_exp(double);
local int  match1(string, string);
local int  isblank(FLOAT *);

void nemo_main()
{
    FITS *fitsfile;
    string fitsname, filter;
    int ndim=3, naxis[3], nx, ny, nz, i, j, k, p,
	idx[MAXSIZE], idy[MAXSIZE], idz[MAXSIZE];
    int nbval=0, plane, bitpix, nblank;
    real bval_out, rmin, rmax, tmp;
    FLOAT *bufin;
    float datamin, datamax, datarange, bscale, bzero, *bufout;
    bool Qshort;
    rproc fie;
    string valid_filters = "linear,sqrt,sqr,log,exp";

    fitsname = getparam("in");
    fitsfile = fitopen(fitsname,"old",ndim,naxis);
    if (fitsfile==NULL) error("Could not open FITS file %s",getparam("in"));
    dprintf(0,"%s: %d * %d * %d\n",
    		getparam("in"),naxis[0],naxis[1],naxis[2]);

    plane = getiparam("plane");
    filter = getparam("filter");
    switch (match1(filter,valid_filters)) {
    case 0:
        fie = filter_lin;       break;
    case 1:
        fie = filter_sqrt;      break;
    case 2:
        fie = filter_sqr;       break;
    case 3:
        fie = filter_log;       break;
    case 4:
        fie = filter_exp;       break;
    default:
        error("Bad filter=%s; valid: %s",filter,valid_filters);
    }
    
    nx = make_idx(getparam("x"),idx,MAXSIZE,naxis[0]);
    ny = make_idx(getparam("y"),idy,MAXSIZE,naxis[1]);
    nz = make_idx(getparam("z"),idz,MAXSIZE,naxis[2]);

    if (nx < 1 || ny < 1 || nz < 1 )
	error("Cannot deal with (%d,%d,%d)",nx,ny,nz);

    if (hasvalue("datamin"))
        datamin=getdparam("datamin");
    else
        if (fitexhd(fitsfile,"DATAMIN"))
            fitrdhdr(fitsfile,"DATAMIN",&datamin,0.0); 
        else
            datamin=getdparam("datamin");
    if (hasvalue("datamax"))
        datamax=getdparam("datamax");
    else
        if (fitexhd(fitsfile,"DATAMAX"))
            fitrdhdr(fitsfile,"DATAMAX",&datamax,0.0); 
        else
            datamax=getdparam("datamax");
#if 0
    datamin = (*fie)((double)datamin);
    datamax = (*fie)((double)datamax);
#endif
    dprintf(0,"Displaying (%d,%d,%d) with %s filter datamin=%g datamax=%g\n",
            nx, ny, nz, filter, datamin, datamax);
    if (datamin >= datamax) error("Cannot display with datamin/max=%g %g",
    	    datamin, datamax);

    /* buffers to read and write */
    bufin  = (FLOAT *) allocate(naxis[0]*sizeof(FLOAT));
    bufout = (float *) allocate(nx*sizeof(float));
    rmin = HUGE;
    rmax = -HUGE;
    nblank = 0;
    look_init(fitsname,nx,ny,datamin,datamax,plane);
    for (k=0; k<nz; k++) {     
        p = idz[k]-1;
        fitsetpl(fitsfile,1,&p);
        for (j=0; j<ny; j++) {      /* loop over all rows */
            fitread(fitsfile,idy[j]-1,bufin);     /* read it from fits file */
            for (i=0; i<nx; i++) {                /* stuff it in memory */
                if (isblank(&bufin[idx[i]-1])) {
                    bufout[i] = bufin[i];
                    nblank++;
                } else {
#if 1
                    if (bufin[idx[i]-1] < datamin) bufin[idx[i]-1] = datamin;
                    if (bufin[idx[i]-1] > datamax) bufin[idx[i]-1] = datamax;
#endif
                    bufout[i] = (*fie)((double)bufin[idx[i]-1]);
                    rmin=MIN(rmin,bufout[i]);
                    rmax=MAX(rmax,bufout[i]);
                }
            }
            nbval += look_loop(j, bufout, nx);
        }
        look_flush();
        if (plane>0) {
	    if (look_record(plane)) {
                dprintf(0,"[Read plane %d and recorded in %d\n",p+1,plane);
                plane++;            
            } else
                dprintf(0,"[Read plane %d but not recorded\n",p+1);
	} else
            dprintf(0,"[Read plane %d]\n",p+1);
    }
    fitclose(fitsfile);

    dprintf(0,"Clipped %d pixels; Found datamin=%g datamax=%g Last REC=%d\n",
            nbval,rmin,rmax,plane-1);
    look_finis();
}

local int make_idx(string axis, int *idx, int maxlen, int axlen)
{
    int i, j, n;
    if (axlen>maxlen)
        warning("Axis %s is too long (%d) to hold %d pixels",axis,axlen,maxlen);
    n = nemoinpi(axis,idx,maxlen);
    if (n>0) {
    	for (i=0, j=0; i<n; i++) {
            if (idx[i]<1 || idx[i]>axlen) continue; /* don't copy: */
            idx[j++] = idx[i];
    	}
    	if (j!=n)
            warning("Only %d/%d elements out of range %s used",j,n,axis);
        return j;
    }
    if (n<0) error("Parsing error %s",axis);
    for (i=0; i<axlen; i++) idx[i] = i+1;
    return axlen;
}

local int match1(string name, string options)
{
    string *sp;
    int i, nsp, retval=-1;
    int len = strlen(name);
    
    sp = burststring(options,",");
    nsp = xstrlen(sp,sizeof(string))-1;
    for (i=0; i<nsp; i++) {
        if (strncmp(sp[i],name,len) == 0) {
            if (retval>=0) {
                warning("%s: matches %s and %s",name,sp[retval],sp[i]);
                return -1;
            } else
                retval = i;
        }
    }
    freestrings(sp);
    return retval;
}

local int blank_val = 0xFFFF;		/* yuck */

local int isblank(FLOAT *bufin)
{
    return memcmp(bufin,&blank_val,sizeof(int))==0;    
}

real filter_lin(double x)  { return x; }
real filter_sqrt(double x) { return sqrt(x); }
real filter_sqr(double x)  { return x*8; }
real filter_exp(double x)  { return exp(x); }
real filter_log(double x)  { return log(x); }


